clients/java/src/main/java/com/google/myanmartools/Phase.java

```diff
@@ -0,0 +1,123 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.myanmartools;
+
+/**
+ * Class implementing a transliteration phase as an array of rules.
+ */
+
+import java.util.ArrayList;
+import java.util.regex.Matcher;
+
+public class Phase {
+
+  private final ArrayList<Rule> phaseRules;
+  private String info;
+  private Boolean debugMode;
+
+  public Phase() {
+    // Initialize the rules for the phase.
+    phaseRules = new ArrayList<>();
+    info = "";
+    debugMode = false;
+  }
+
+  public void setInfo(String newInfo) {
+    info = newInfo;
+  }
+
+  public void setDebugMode(Boolean newMode) {
+    debugMode = newMode;
+  }
+
+  public void addRule(Rule newRule) {
+    phaseRules.add(newRule);
+    // Put the rule number within the phase.
+    newRule.setInfo("" + phaseRules.size());
+  }
+
+  public String printPhase() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("  Phase ").append(info)
+        .append(" has ").
+        append(phaseRules.size()).
+        append(" rules\n");
+
+    for (Rule rule : phaseRules) {
+      sb.append(rule.printRule());
+    }
+    return sb.toString();
+  }
+
+  public String runPhase(Phase phase, String inString) {
+    // Run all the rules of this phase.
+    String outString = "";
+    String midString = inString;
+    Boolean startOfString = true;
+
+    if (debugMode) {
+      System.out.println("Phase " + info + ", input= " + inString + "  (" + inString.length() + ")");
+    }
+
+    while (midString.length() > 0) {
+      // Move through the string, matching / applying rules .
+      Boolean foundRule = false;
+
+      for (Rule rule : phaseRules) {
+        // Check if the rule applies.
+        if (!rule.matchOnStart || startOfString) {
+          Matcher m = rule.pattern.matcher(midString);
+          if (m.find()) {
+
+            if (debugMode) {
+              // Debug
+              System.out.println("  Matched rule " + rule.info + " = " + rule.pattern + " --> " +
+                  rule.substitution);
+              System.out.println("    m.group(0):  " + m.group(0) + " (" + m.group(0).length() +")");
+            }
+            foundRule = true;
+            int rightPartSize = midString.length() - m.group(0).length();
+
+
+            midString = m.replaceFirst(rule.substitution);
+
+            if (rule.revisitPosition < 0) {
+              // Reset the new position to the end of the subsitution.
+              int newStart = midString.length() - rightPartSize;
+              outString = outString + midString.substring(0, newStart);
```

Comment:
Is there an easy way to make `outString` and/or `midString`, StringBuilders instead of Strings?  This just looks very inefficient.

Reply:
outString now build using StringBuilder.